### PR TITLE
enable api getUsers for subadmins

### PR DESCRIPTION
**Pull request by michag86**
Make it possible to use getUsers from provisioning api for subadmins.

### CODE REVIEW

File: apps/provisioning_api/lib/users.php [hunk 1]
```diff
@@ -68,7 +69,25 @@ public function getUsers(){
...
-		$users = $this->userManager->search($search, $limit, $offset);
+		$user = $this->userSession->getUser();
+
+		// Admin? Or SubAdmin?
+		if($this->groupManager->isAdmin($user->getUID())){
+			$users = $this->userManager->search($search, $limit, $offset);
+		} else {
+			$subAdminOfGroups = \OC_SubAdmin::getSubAdminsGroups($user->getUID());
+
+			if($limit === null) { $limit = -1; }
```
Comment (MorrisJobke) [inline apps/provisioning_api/lib/users.php]: please put the statement on a new line.

File: apps/provisioning_api/lib/users.php [hunk 2]
```diff
@@ -68,7 +69,25 @@ public function getUsers(){
...
+			if($offset === null) { $offset = 0; }
+			$batch = [];
+
+			$groupUsers = OC_Group::displayNamesInGroups($subAdminOfGroups, $search, $limit, $offset);
```
Comment (MorrisJobke) [inline apps/provisioning_api/lib/users.php]: @rullzer @Xenopathic @nickvergessen This method is missing in the public namespace. Also the private implementation isn't that good to put into public namespace (see TODO). Any ideas?

**Comment by MorrisJobke** [comment]
Beside my comments this looks good.

**Comment by MorrisJobke** [comment]
cc @DeepDiver1975 Provisioning API + SubAdmin :question:

**Comment by MorrisJobke** [comment]
cc @tomneedham

File: apps/provisioning_api/lib/users.php [hunk 3]
```diff
@@ -68,7 +69,29 @@ public function getUsers(){
...
+
+			if($limit === null) { 
+				$limit = -1; 
+			}
+			if($offset === null) {
+				$offset = 0; 
+			}
+			$batch = [];
+
+			$groupUsers = OC_Group::displayNamesInGroups($subAdminOfGroups, $search, $limit, $offset);
```
Comment (rullzer) [inline apps/provisioning_api/lib/users.php]: No... I just spend some time to get rid of most of the OC_* calls in the provisioning api.

Just use a simple loop and use ```$this->groupManager->displayNamesInGroups```

**Comment by rullzer** [comment]
Please fix the test case
And add new test cases

**Comment by rullzer** [comment]
Addition does make a lot of sense! Thanks @michag86.

File: apps/provisioning_api/lib/users.php (hunk 3 above)
Comment (michag86) [inline apps/provisioning_api/lib/users.php]: But if I use this in a loop than limit and offset are used as expected. If the subadmin uses limit=2 and has two groups to manage he gets 4 results.

**Comment by MorrisJobke** [comment]
@rullzer The OC_Subadmin stuff isn't available in any proper implementation :cry:

**Comment by rullzer** [comment]
> @rullzer The OC_Subadmin stuff isn't available in any proper implementation :cry: 

Well that is known. But the provisioning API is already using the calls currently. So I do not really see an issue with that. It is far from ideal, I admit that.

**Comment by rullzer** [comment]
Rebased and squashed.

**Comment by rullzer** [comment]
Issue for displayNamesInGroups in https://github.com/owncloud/core/issues/18418

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by rullzer** [comment]
So oracle is fine now.

Let the reviews begin!

File: apps/provisioning_api/lib/users.php [hunk 4]
```diff
@@ -68,7 +69,31 @@ public function getUsers(){
 		$limit = !empty($_GET['limit']) ? $_GET['limit'] : null;
 		$offset = !empty($_GET['offset']) ? $_GET['offset'] : null;
 
-		$users = $this->userManager->search($search, $limit, $offset);
+		// Check if user is logged in
+		$user = $this->userSession->getUser();
```
Comment (MorrisJobke) [inline apps/provisioning_api/lib/users.php]: Isn't this already done by the `API::SUBADMIN_AUTH`?

**Comment by MorrisJobke** [comment]
Tested and works :+1:

CURL command: `curl http://test:test@localhost/master/ocs/v1.php/cloud/users`

**Comment by PVince81** [comment]
We're already past feature freeze unfortunately.

@MTRichards @cmonteroluque can we get this useful addition get scheduled for 9.0 ?

**Comment by MTRichards** [comment]
Will link it for consideration

**Comment by PVince81** [comment]
CC @SergioBertolinSG for additional provisioning API API-level tests

**Comment by PVince81** [comment]
Tested, works :+1:

**Comment by DeepDiver1975** [comment]
@SergioBertolinSG please add an integration test to test this behavior - THX

**Pull request merged**